File: app/routers/whitelist.py
import os
from datetime import timezone
from fastapi import APIRouter, Depends, Request, HTTPException, Query, Response
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete, func
from app.database import get_db
from app.models import WhitelistAddress, WhitelistState
from app.schemas import AddressIn, WhitelistResponse, WhitelistMeta, BulkWriteResponse, DeleteResponse, MerkleRootResponse, EntryIn, EntriesResponse, WhitelistEntry, SetMerkleRootIn
from app.merkle import build_merkle_tree
from app.auth import get_session_address
from app.rate_limit import limiter

router = APIRouter(prefix="/api/whitelist", tags=["whitelist"])

ADMIN_ADDRESS = os.environ.get("ADMIN_ADDRESS", "").lower()

def _require_admin(request: Request):
    addr = get_session_address(request)
    if not addr:
        raise HTTPException(status_code=401, detail="Unauthorized")
    if addr.lower() != ADMIN_ADDRESS:
        raise HTTPException(status_code=403, detail="Forbidden")

async def _recalc_merkle(db: AsyncSession) -> str:
    # Flush any pending changes so the SELECT reads current data
    await db.flush()
    state = (await db.execute(select(WhitelistState).where(WhitelistState.id == 1))).scalar_one_or_none()
    # Respect manual override — admin-set roots survive whitelist edits until
    # the override is cleared via DELETE /api/whitelist/merkle-root.
    if state and state.manual_override:
        await db.commit()
        return state.merkle_root
    result = await db.execute(select(WhitelistAddress.address).order_by(WhitelistAddress.id))
    addresses = [row[0] for row in result.all()]
    root = build_merkle_tree(addresses)["root"] if addresses else "0x0"
    if state:
        state.merkle_root = root
        await db.flush()
        await db.refresh(state)
    else:
        db.add(WhitelistState(id=1, merkle_root=root))
    await db.commit()
    return root

@router.get("", response_model=WhitelistResponse)
@limiter.limit("100/minute")
async def get_whitelist(
    request: Request,
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar()

    result = await db.execute(
        select(WhitelistAddress.address).order_by(WhitelistAddress.id).offset(offset).limit(limit)
    )
    addresses = [row[0] for row in result.all()]

    state = await db.get(WhitelistState, 1)
    return WhitelistResponse(
        addresses=addresses,
        total=total,
        limit=limit,
        offset=offset,
        has_more=offset + limit < total,
        metadata=WhitelistMeta(
            merkle_root=state.merkle_root if state else "0x0",
            last_updated=state.last_updated.isoformat() if state else "",
            timestamp=int(__import__("time").time() * 1000),
            manual_override=bool(state.manual_override) if state else False,
        ),
    )

@router.post("", response_model=BulkWriteResponse)
@limiter.limit("10/minute")
async def bulk_write_whitelist(
    request: Request,
    body: AddressIn,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    if body.action == "replace":
        await db.execute(delete(WhitelistAddress))
        await db.commit()

    existing_result = await db.execute(select(WhitelistAddress.address_lower))
    existing_set = {row[0] for row in existing_result.all()}

    to_add = [a for a in body.addresses if a.lower() not in existing_set]
    for addr in to_add:
        db.add(WhitelistAddress(address=addr, address_lower=addr.lower()))
    await db.commit()

    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar()

    root = await _recalc_merkle(db)
    skipped = len(body.addresses) - len(to_add)
    return BulkWriteResponse(success=True, count=total, added=len(to_add), skipped=skipped, duplicates=skipped, merkle_root=root)

@router.post("/add", response_model=BulkWriteResponse)
@limiter.limit("10/minute")
async def add_addresses(
    request: Request,
    body: AddressIn,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    body.action = "add"
    return await bulk_write_whitelist(request, body, db)


@router.post("/entry", response_model=BulkWriteResponse)
@limiter.limit("10/minute")
async def add_entry(
    request: Request,
    body: EntryIn,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    """Add a single (address, name) entry. If the address already exists, the
    name is updated rather than duplicated."""
    existing = await db.execute(
        select(WhitelistAddress).where(WhitelistAddress.address_lower == body.address.lower())
    )
    row = existing.scalar_one_or_none()
    if row is None:
        db.add(WhitelistAddress(address=body.address, address_lower=body.address.lower(), name=body.name))
        added = 1
        skipped = 0
    else:
        # Address already on the list — only update the name if it was missing.
        if not row.name:
            row.name = body.name
        added = 0
        skipped = 1
    await db.commit()

    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar()
    root = await _recalc_merkle(db)
    return BulkWriteResponse(
        success=True, count=total, added=added, skipped=skipped, duplicates=skipped, merkle_root=root
    )


@router.get("/entries", response_model=EntriesResponse)
@limiter.limit("100/minute")
async def get_entries(
    request: Request,
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    """Admin-only listing that includes name + added_at alongside the address."""
    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar() or 0

    result = await db.execute(
        select(WhitelistAddress).order_by(WhitelistAddress.id).limit(limit).offset(offset)
    )
    rows = result.scalars().all()
    entries = [
        WhitelistEntry(
            address=r.address,
            name=r.name,
            added_at=r.added_at.isoformat() if r.added_at else None,
        )
        for r in rows
    ]

    state = (await db.execute(select(WhitelistState).where(WhitelistState.id == 1))).scalar_one_or_none()
    return EntriesResponse(
        entries=entries,
        total=total,
        limit=limit,
        offset=offset,
        has_more=offset + limit < total,
        metadata=WhitelistMeta(
            merkle_root=state.merkle_root if state else "0x0",
            last_updated=state.last_updated.isoformat() if state else "",
            timestamp=int(__import__("time").time() * 1000),
            manual_override=bool(state.manual_override) if state else False,
        ),
    )

@router.delete("/{address}", response_model=DeleteResponse)
@limiter.limit("5/minute")
async def delete_address(
    address: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    import re
    if not re.match(r"^0x[a-fA-F0-9]{40}$", address):
        raise HTTPException(status_code=400, detail="Invalid address format")

    result = await db.execute(
        select(WhitelistAddress).where(WhitelistAddress.address_lower == address.lower())
    )
    row = result.scalar_one_or_none()
    if not row:
        raise HTTPException(status_code=404, detail="Address not found")

    await db.delete(row)
    await db.commit()

    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar()
    root = await _recalc_merkle(db)
    return DeleteResponse(success=True, count=total, removed=address, merkle_root=root)

@router.get("/export")
@limiter.limit("100/minute")
async def export_whitelist(
    request: Request,
    format: str = Query("json", regex="^(json|csv|txt)$"),
    include_merkle_root: bool = Query(True),
    db: AsyncSession = Depends(get_db),
):
    import json as json_mod
    from datetime import datetime
    result = await db.execute(select(WhitelistAddress.address).order_by(WhitelistAddress.id))
    addresses = [row[0] for row in result.all()]
    state = await db.get(WhitelistState, 1)
    date_str = datetime.now().strftime("%Y-%m-%d")

    if format == "csv":
        content = "address\n" + "\n".join(addresses)
        return Response(content=content, media_type="text/csv",
                        headers={"Content-Disposition": f'attachment; filename="whitelist-{date_str}.csv"'})
    elif format == "txt":
        content = "\n".join(addresses)
        return Response(content=content, media_type="text/plain",
                        headers={"Content-Disposition": f'attachment; filename="whitelist-{date_str}.txt"'})
    else:
        output = {"whitelist": addresses}
        if include_merkle_root:
            output["metadata"] = {
                "total": len(addresses),
                "merkle_root": state.merkle_root if state else None,
                "exported_at": datetime.now().isoformat(),
            }
        content = json_mod.dumps(output, indent=2)
        return Response(content=content, media_type="application/json",
                        headers={"Content-Disposition": f'attachment; filename="whitelist-{date_str}.json"'})

@router.get("/merkle-root", response_model=MerkleRootResponse)
@limiter.limit("100/minute")
async def get_merkle_root(request: Request, db: AsyncSession = Depends(get_db)):
    state = await db.get(WhitelistState, 1)
    total_result = await db.execute(select(func.count()).select_from(WhitelistAddress))
    total = total_result.scalar()
    return MerkleRootResponse(
        root=state.merkle_root if state else "0x0",
        count=total,
        generated_at=state.last_updated.isoformat() if state else "",
        manual_override=bool(state.manual_override) if state else False,
    )

@router.put("/merkle-root", response_model=MerkleRootResponse)
@limiter.limit("10/minute")
async def set_merkle_root(
    request: Request,
    body: SetMerkleRootIn,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    # Sets manual_override=true so subsequent whitelist edits do not overwrite
    # this value via _recalc_merkle. Clear with DELETE /api/whitelist/merkle-root.
    state = await db.get(WhitelistState, 1)
    if state:
        state.merkle_root = body.root
        state.manual_override = True
    else:
        db.add(WhitelistState(id=1, merkle_root=body.root, manual_override=True))
    await db.commit()
    state = await db.get(WhitelistState, 1)
    total = (await db.execute(select(func.count()).select_from(WhitelistAddress))).scalar()
    return MerkleRootResponse(
        root=state.merkle_root,
        count=total,
        generated_at=state.last_updated.isoformat() if state else "",
        manual_override=bool(state.manual_override) if state else False,
    )

@router.delete("/merkle-root", response_model=MerkleRootResponse)
@limiter.limit("10/minute")
async def clear_merkle_root_override(
    request: Request,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_require_admin),
):
    # Clears manual_override and recalculates from the current whitelist.
    state = await db.get(WhitelistState, 1)
    if state:
        state.manual_override = False
    await db.commit()
    root = await _recalc_merkle(db)
    state = await db.get(WhitelistState, 1)
    total = (await db.execute(select(func.count()).select_from(WhitelistAddress))).scalar()
    return MerkleRootResponse(
        root=root,
        count=total,
        generated_at=state.last_updated.isoformat() if state else "",
        manual_override=bool(state.manual_override) if state else False,
    )

@router.post("/test")
@limiter.limit("10/minute")
async def test_address(request: Request, body: dict, db: AsyncSession = Depends(get_db)):
    import re
    from datetime import datetime
    address = body.get("address", "")
    if not re.match(r"^0x[a-fA-F0-9]{40}$", address):
        raise HTTPException(status_code=400, detail="Invalid address format")

    result = await db.execute(select(WhitelistAddress.address).order_by(WhitelistAddress.id))
    addresses = [row[0] for row in result.all()]

    if not addresses:
        return {"is_whitelisted": False, "address": address, "proof": [], "root": "0x0", "leaf_index": None, "generated_at": datetime.now().isoformat()}

    tree = build_merkle_tree(addresses)
    lower = address.lower()
    lower_list = [a.lower() for a in addresses]
    leaf_index = lower_list.index(lower) if lower in lower_list else None

    from app.merkle import get_proof
    proof = get_proof(tree, address) if leaf_index is not None else []

    return {"is_whitelisted": leaf_index is not None, "address": address, "proof": proof, "root": tree["root"], "leaf_index": leaf_index, "generated_at": datetime.now().isoformat()}

File: src/routes/whitelist.ts
import { Router, Request, Response, NextFunction } from "express";
import rateLimit from "express-rate-limit";
import { PoolClient } from "pg";
import pool from "../pool";
import { buildMerkleTree, getProof } from "../merkle";
import { verifySessionCookie } from "../walletAuth";
import { HttpError } from "../errors";

const router = Router();
const ETH_ADDRESS_RE = /^0x[a-fA-F0-9]{40}$/;
const MERKLE_ROOT_RE = /^0x[a-fA-F0-9]{64}$/;
const ADMIN_ADDRESS = (process.env.ADMIN_ADDRESS ?? "").toLowerCase();

const readLimit   = rateLimit({ windowMs: 60_000, limit: 100, standardHeaders: "draft-7", legacyHeaders: false });
const writeLimit  = rateLimit({ windowMs: 60_000, limit: 10,  standardHeaders: "draft-7", legacyHeaders: false });
const deleteLimit = rateLimit({ windowMs: 60_000, limit: 5,   standardHeaders: "draft-7", legacyHeaders: false });

function requireAdmin(req: Request): void {
  const addr = verifySessionCookie(req.headers.cookie);
  if (!addr) throw new HttpError(401, "Unauthorized");
  if (addr.toLowerCase() !== ADMIN_ADDRESS) throw new HttpError(403, "Forbidden");
}

async function recalcMerkle(client: PoolClient): Promise<string> {
  const { rows: [state] } = await client.query("SELECT * FROM whitelist_state_get()");
  if (state?.manual_override) return state.merkle_root as string;
  const { rows: addrRows } = await client.query("SELECT * FROM whitelist_addresses_all()");
  const addresses = addrRows.map((r: { address: string }) => r.address);
  const root = addresses.length ? buildMerkleTree(addresses).root : "0x0";
  await client.query("SELECT whitelist_state_update_root($1)", [root]);
  return root;
}

async function bulkWrite(res: Response, next: NextFunction, addresses: string[], action: "add" | "replace") {
  if (!Array.isArray(addresses) || !addresses.length) {
    res.status(422).json({ detail: "addresses must be non-empty" }); return;
  }
  if (addresses.length > 1000) {
    res.status(422).json({ detail: "maximum 1000 addresses per request" }); return;
  }
  const invalid = addresses.filter(a => !ETH_ADDRESS_RE.test(a));
  if (invalid.length) {
    res.status(422).json({ detail: `invalid addresses: ${invalid.join(", ")}` }); return;
  }
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    if (action === "replace") await client.query("SELECT whitelist_clear()");
    const { rows: existingRows } = await client.query("SELECT * FROM whitelist_existing_lowers()");
    const existingSet = new Set(existingRows.map((r: { address_lower: string }) => r.address_lower));
    const toAdd = addresses.filter(a => !existingSet.has(a.toLowerCase()));
    for (const addr of toAdd) {
      await client.query("SELECT whitelist_add($1, $2)", [addr, addr.toLowerCase()]);
    }
    const root = await recalcMerkle(client);
    await client.query("COMMIT");
    const { rows: [{ count }] } = await client.query("SELECT whitelist_count() AS count");
    const skipped = addresses.length - toAdd.length;
    res.json({ success: true, count: Number(count), added: toAdd.length, skipped, duplicates: skipped, merkle_root: root });
  } catch (e) {
    await client.query("ROLLBACK");
    next(e);
  } finally {
    client.release();
  }
}

// ── Specific paths first (before /:address) ──────────────────────────────────

// GET /api/whitelist/entries — admin-only, includes name + added_at
router.get("/entries", readLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
    const limit  = Math.min(Number(req.query.limit  ?? 100), 1000);
    const offset = Number(req.query.offset ?? 0);
    const [entriesRes, countRes, stateRes] = await Promise.all([
      pool.query("SELECT * FROM whitelist_entries($1, $2)", [limit, offset]),
      pool.query("SELECT whitelist_count() AS count"),
      pool.query("SELECT * FROM whitelist_state_get()"),
    ]);
    const total = Number(countRes.rows[0].count);
    const state = stateRes.rows[0];
    res.json({
      entries: entriesRes.rows.map((r: { id: string; address: string; user_id: string | null; is_whitelisted: boolean; added_at: Date | null }) => ({
        id: r.id,
        address: r.address,
        user_id: r.user_id ?? null,
        is_whitelisted: r.is_whitelisted,
        added_at: r.added_at?.toISOString() ?? null,
      })),
      total, limit, offset,
      has_more: offset + limit < total,
      metadata: {
        merkle_root: state?.merkle_root ?? "0x0",
        last_updated: state?.last_updated?.toISOString() ?? "",
        timestamp: Date.now(),
        manual_override: Boolean(state?.manual_override),
      },
    });
  } catch (e) { next(e); }
});

// GET /api/whitelist/export — public
router.get("/export", readLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const format           = (req.query.format as string) ?? "json";
    const includeMerkleRoot = req.query.include_merkle_root !== "false";
    const [addrRes, stateRes] = await Promise.all([
      pool.query("SELECT * FROM whitelist_addresses_all()"),
      pool.query("SELECT * FROM whitelist_state_get()"),
    ]);
    const addresses = addrRes.rows.map((r: { address: string }) => r.address);
    const state = stateRes.rows[0];
    const dateStr = new Date().toISOString().slice(0, 10);

    if (format === "csv") {
      res.setHeader("Content-Type", "text/csv");
      res.setHeader("Content-Disposition", `attachment; filename="whitelist-${dateStr}.csv"`);
      res.send("address\n" + addresses.join("\n"));
    } else if (format === "txt") {
      res.setHeader("Content-Type", "text/plain");
      res.setHeader("Content-Disposition", `attachment; filename="whitelist-${dateStr}.txt"`);
      res.send(addresses.join("\n"));
    } else {
      const output: Record<string, unknown> = { whitelist: addresses };
      if (includeMerkleRoot) {
        output.metadata = {
          total: addresses.length,
          merkle_root: state?.merkle_root ?? null,
          exported_at: new Date().toISOString(),
        };
      }
      res.setHeader("Content-Type", "application/json");
      res.setHeader("Content-Disposition", `attachment; filename="whitelist-${dateStr}.json"`);
      res.send(JSON.stringify(output, null, 2));
    }
  } catch (e) { next(e); }
});

// GET /api/whitelist/merkle-root — public
router.get("/merkle-root", readLimit, async (_req: Request, res: Response, next: NextFunction) => {
  try {
    const [stateRes, countRes] = await Promise.all([
      pool.query("SELECT * FROM whitelist_state_get()"),
      pool.query("SELECT whitelist_count() AS count"),
    ]);
    const state = stateRes.rows[0];
    res.json({
      root: state?.merkle_root ?? "0x0",
      count: Number(countRes.rows[0].count),
      generated_at: state?.last_updated?.toISOString() ?? "",
      manual_override: Boolean(state?.manual_override),
    });
  } catch (e) { next(e); }
});

// PUT /api/whitelist/merkle-root — admin-only, sets manual override
router.put("/merkle-root", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
    const { root } = (req.body ?? {}) as { root?: string };
    if (!root || !MERKLE_ROOT_RE.test(root)) {
      res.status(422).json({ detail: "root must be a 0x-prefixed 32-byte hex string" }); return;
    }
    await pool.query("SELECT whitelist_state_set_override($1)", [root]);
    const [stateRes, countRes] = await Promise.all([
      pool.query("SELECT * FROM whitelist_state_get()"),
      pool.query("SELECT whitelist_count() AS count"),
    ]);
    const state = stateRes.rows[0];
    res.json({
      root: state.merkle_root,
      count: Number(countRes.rows[0].count),
      generated_at: state.last_updated?.toISOString() ?? "",
      manual_override: true,
    });
  } catch (e) { next(e); }
});

// DELETE /api/whitelist/merkle-root — admin-only, clears manual override
router.delete("/merkle-root", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
  } catch (e) { next(e); return; }
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    await client.query("SELECT whitelist_state_clear_override()");
    const root = await recalcMerkle(client);
    await client.query("COMMIT");
    const [stateRes, countRes] = await Promise.all([
      pool.query("SELECT * FROM whitelist_state_get()"),
      pool.query("SELECT whitelist_count() AS count"),
    ]);
    const state = stateRes.rows[0];
    res.json({
      root,
      count: Number(countRes.rows[0].count),
      generated_at: state?.last_updated?.toISOString() ?? "",
      manual_override: false,
    });
  } catch (e) {
    await client.query("ROLLBACK");
    next(e);
  } finally {
    client.release();
  }
});

// POST /api/whitelist/add — add (no replace)
router.post("/add", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
  } catch (e) { next(e); return; }
  const { addresses } = (req.body ?? {}) as { addresses?: string[] };
  await bulkWrite(res, next, addresses ?? [], "add");
});

// POST /api/whitelist/entry — add single entry with optional user_id / is_whitelisted
router.post("/entry", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
    const { address, user_id, is_whitelisted } = (req.body ?? {}) as {
      address?: string;
      user_id?: string | null;
      is_whitelisted?: boolean;
    };
    if (!address || !ETH_ADDRESS_RE.test(address)) {
      res.status(422).json({ detail: "invalid address" }); return;
    }
    const whitelisted = is_whitelisted !== false; // default true
    const client = await pool.connect();
    try {
      await client.query("BEGIN");
      const { rows: [existing] } = await client.query(
        "SELECT * FROM whitelist_entry_get($1)",
        [address.toLowerCase()]
      );
      let added   = 0;
      let skipped = 0;
      if (!existing) {
        await client.query(
          "SELECT whitelist_add($1, $2, $3, $4)",
          [address, address.toLowerCase(), user_id ?? null, whitelisted]
        );
        added = 1;
      } else {
        skipped = 1;
      }
      const root = await recalcMerkle(client);
      await client.query("COMMIT");
      const { rows: [{ count }] } = await client.query("SELECT whitelist_count() AS count");
      res.json({ success: true, count: Number(count), added, skipped, duplicates: skipped, merkle_root: root });
    } catch (e) {
      await client.query("ROLLBACK");
      throw e;
    } finally {
      client.release();
    }
  } catch (e) { next(e); }
});

// POST /api/whitelist/test — public, test if address is whitelisted
router.post("/test", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { address } = (req.body ?? {}) as { address?: string };
    if (!address || !ETH_ADDRESS_RE.test(address)) {
      res.status(400).json({ detail: "Invalid address format" }); return;
    }
    const { rows } = await pool.query("SELECT * FROM whitelist_addresses_all()");
    const addresses = rows.map((r: { address: string }) => r.address);
    if (!addresses.length) {
      res.json({ is_whitelisted: false, address, proof: [], root: "0x0", leaf_index: null, generated_at: new Date().toISOString() });
      return;
    }
    const tree     = buildMerkleTree(addresses);
    const lower    = address.toLowerCase();
    const lowerList = addresses.map((a: string) => a.toLowerCase());
    const leafIndex = lowerList.indexOf(lower);
    const isWhitelisted = leafIndex !== -1;
    const proof = isWhitelisted ? getProof(tree, address) : [];
    res.json({ is_whitelisted: isWhitelisted, address, proof, root: tree.root, leaf_index: isWhitelisted ? leafIndex : null, generated_at: new Date().toISOString() });
  } catch (e) { next(e); }
});

// DELETE /api/whitelist/:address — admin-only
router.delete("/:address", deleteLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
  } catch (e) { next(e); return; }
  const { address } = req.params;
  if (!ETH_ADDRESS_RE.test(address)) {
    res.status(400).json({ detail: "Invalid address format" }); return;
  }
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    // whitelist_remove throws P0002 if not found — errorHandler maps it to 404
    await client.query("SELECT whitelist_remove($1)", [address.toLowerCase()]);
    const root = await recalcMerkle(client);
    await client.query("COMMIT");
    const { rows: [{ count }] } = await client.query("SELECT whitelist_count() AS count");
    res.json({ success: true, count: Number(count), removed: address, merkle_root: root });
  } catch (e) {
    await client.query("ROLLBACK");
    next(e);
  } finally {
    client.release();
  }
});

// ── Root routes ───────────────────────────────────────────────────────────────

// GET /api/whitelist
router.get("/", readLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
    const limit  = Math.min(Number(req.query.limit  ?? 100), 1000);
    const offset = Number(req.query.offset ?? 0);
    const [countRes, addrRes, stateRes] = await Promise.all([
      pool.query("SELECT whitelist_count() AS count"),
      pool.query("SELECT * FROM whitelist_list($1, $2)", [limit, offset]),
      pool.query("SELECT * FROM whitelist_state_get()"),
    ]);
    const total = Number(countRes.rows[0].count);
    const state = stateRes.rows[0];
    res.json({
      addresses: addrRes.rows.map((r: { address: string }) => r.address),
      total, limit, offset,
      has_more: offset + limit < total,
      metadata: {
        merkle_root: state?.merkle_root ?? "0x0",
        last_updated: state?.last_updated?.toISOString() ?? "",
        timestamp: Date.now(),
        manual_override: Boolean(state?.manual_override),
      },
    });
  } catch (e) { next(e); }
});

// POST /api/whitelist — bulk write (add or replace)
router.post("/", writeLimit, async (req: Request, res: Response, next: NextFunction) => {
  try {
    requireAdmin(req);
  } catch (e) { next(e); return; }
  const { addresses, action = "add" } = (req.body ?? {}) as { addresses?: string[]; action?: string };
  await bulkWrite(res, next, addresses ?? [], (action === "replace" ? "replace" : "add"));
});

export default router;

export { recalcMerkle };
